snapcraft/internal/deltas/_deltas.py

```diff
@@ -0,0 +1,210 @@
+        stdout_file = open(stdout_path, 'wb')
+
+        stderr_path = self.find_unique_file_name(
+            os.path.join(workdir, '{}-err'.format(self.delta_format)))
+        stderr_file = open(stderr_path, 'wb')
+
+        return workdir, stdout_path, stdout_file, stderr_path, stderr_file
+
+    def _update_progress_indicator(self, proc, progress_indicator):
+        """update the progress indicator"""
+        # the caller should start the progressbar outside
+        ret = None
+        count = 0
+        ret = proc.poll()
+        while ret is None:
+            if count >= progress_indicator.maxval:
+                progress_indicator.start()
+                count = 0
+            progress_indicator.update(count)
+            count += 1
+            time.sleep(.2)
+            ret = proc.poll()
+        print('')
+        # the caller should finish the progressbar outside
+
+    def make_delta(self, output_dir=None, progress_indicator=None):
+        """call the delta generation tool to create the delta file.
+
+        returns: generated delta file path
+        """
+        logger.info('Generating {} delta for {}->{}.'.format(
+                self.delta_format, self.source_path, self.target_path))
+
+        if output_dir is not None:
+            # consider creating the delta file in the specified output_dir
+            # with generated filename.
+            if not os.path.exists(output_dir):
+                os.makedirs(output_dir, exist_ok=True)
+
+            _, _file_name = os.path.split(self.target_path)
+            full_filename = os.path.join(output_dir, _file_name)
+            delta_file = self.find_unique_file_name(
+                '{}.{}'.format(full_filename, self.delta_file_extname))
+        else:
+            # create the delta file under the target_path with
+            # the generated filename.
+            delta_file = self.find_unique_file_name(
+                '{}.{}'.format(self.target_path, self.delta_file_extname))
+
+        delta_cmd = self.get_delta_cmd(self.source_path,
+                                       self.target_path,
+                                       delta_file)
+
+        workdir, stdout_path, stdout_file, \
+            stderr_path, stderr_file = self._setup_std_output(self.source_path)
+
+        proc = subprocess.Popen(
+            delta_cmd,
+            stdout=stdout_file,
+            stderr=stderr_file,
+            cwd=workdir
+        )
+
+        if progress_indicator is not None and isinstance(progress_indicator,
+                                                         ProgressBar):
+            self._update_progress_indicator(proc, progress_indicator)
+        else:
+            proc.wait()
+
+        stdout_file.close()
+        stderr_file.close()
+
+        # Success is exiting with 0 or 1. Yes, really. I know.
+        # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=212189
+        if proc.returncode not in (0, 1):
```

Comment:
Only xdelta returns 1 (in the case where src and target are the same, so the output delta is empty), yet it's being checked in the base class for all current and future delta algorithms.